keps/sig-api-machinery/20180415-crds-to-ga.md

```diff
@@ -0,0 +1,256 @@
+* if a missing feature forces or encourages users to implement non-Kubernetes-like APIs, and therefore damages the ecosystem long term, it belongs on this list.
+* If a feature cannot be added as an after-thought of a GA API, it or some preparation belongs on this list.
+
+
+### Non-Goals
+
+Full feature parity with native kubernetes resource types is not a GA graduation goal. See above guiding principles and below Post-GA tasks.
+
+## Proposal
+
+Objectives to graduate CRDs to GA, each of which is described in more detail further below:
+
+* Defaulting and pruning for custom resources is implemented
+* CRD v1 schemas are restricted to a subset of the OpenAPI specification (and there is a v1beta1 compatibility plan)
+* Generator exists for CRD Validation Schema v3 (Kubebuilder)
+* CustomResourceWebhookConversion API is GA ready
+* CustomResourceSubresources API is GA ready
+
+Bug fixes required to graduate CRDs to GA:
+
+* See “Required for GA” issues tracked via the [CRD Project Board](https://github.com/orgs/kubernetes/projects/28).
+
+For additional details on already completed features, see the [Umbrella Issue](https://github.com/kubernetes/kubernetes/issues/58682).
+
+See [Post-GA tasks](#post-ga-tasks) for decided out-of-scope features.
+
+### Defaulting and pruning for custom resources is implemented
+
+Both defaulting and pruning and also read-only validation a blocked by the
+OpenAPI subset definition (next point). An update of the [old Pruning for
+CustomResources KEP](kubernetes/enhancements#709) and the implementation
+([pruning PR](kubernetes/kubernetes#64558), [defaulting
+PR](kubernetes/kubernetes#63604)), are follow-ups as soon as unblocked.
+
+### CRD v1 schemas are restricted to a subset of the OpenAPI specification
+
+See [Vanilla OpenAPI Subset Design](https://docs.google.com/document/d/1pcGlbmw-2Y0JJs9hsYnSBXamgG9TfWtHY6eh80zSTd8)
+
+### Generator exists for CRD Validation Schema v3 (Kubebuilder)
+
+See [kubebuilder’s
+crd-gen](https://github.com/kubernetes-sigs/controller-tools/tree/master/cmd/crd)
+and a more general
+[crd-schema-gen](https://github.com/kubernetes-sigs/controller-tools/pull/183),
+to be integrated into kubebuidler’s controller-tools.
+
+### CustomResourceWebhookConversion API is GA ready
+
+Currently conversion webhooks are v1alpha1. Our goal is to move to v1beta1 in
+the Kubernetes 1.15 release, and then graduate this to GA as part of the CRD to
+GA graduation; we won't track this as a separate feature from CRDs once it's in
+v1beta1.
+
+Conversion webhooks to beta tasks:
+- ConversionReview API v1 (incl. fields that need to be updated for v1beta1 or v1)
+- Add conversion webhook tests enumerated in below [Test Plan](#test-plan)
+* Add comprehensive metrics to conversion webhooks
+
+### CustomResourceSubresources API is GA ready
+
+Currently custom resource subresources are v1beta1 and provide support for the
+/status and /scale subresources.  We plan to move this to v1beta1 in Kubernetes
```

Comment:
drop "move this to v1beta1 in 1.15 release"

Reply:
Removed.